Add error to specs for undocumented analytics event params (LG-5969)

zachmargolis opened the pull request:
## 🎫 Ticket

[LG-5969](https://cm-jira.usa.gov/browse/LG-5969)


## 🛠 Summary of changes

Adds code to the test suite that will error when undocumented params are sent to the `track_event` method.

### Spec changes

There are two kinds of exceptions right now:

1. We seem to intentionally log extra undocumented params to make sure they get passed through
2. We have analytics methods that don't take params but do *add* extra stuff to the payload as new undocumented keys like so: https://github.com/18F/identity-idp/blob/84dcd52a872ff429babbb6d5c75ff40f162771a9/app/services/analytics_events.rb#L948-L953

So to get around that, we have some new spec metadata that can be used to annotate:

```ruby
context 'some spec', allowed_extra_analytics: [:my_param_name] do
  it 'logs the stuff' do
    analytics.track_some_event(my_param_name: true)
  end
end
```

aduth commented on spec/services/doc_auth_router_spec.rb:
```diff
@@ -74,7 +74,7 @@ def reload_ab_test_initializer!
       let(:discriminator) { nil }
 
-      it 'is the default vendor, and logs analytics events' do
+      it 'is the default vendor, and logs analytics events', allowed_extra_analytics: [:error] do
```
Should this be allowed? 🤔

zachmargolis replied:
I think no, this is what I was [referencing in the PR description](https://github.com/18F/identity-idp/pull/9946#:~:text=We%20have%20analytics%20methods%20that%20don%27t%20take%20params%20but%20do%20add%20extra%20stuff%20to%20the%20payload%20as%20new%20undocumented%20keys%20like%20so%3A), where the `idv_doc_auth_randomizer_defaulted` takes no params (so there's nothing to document) but adds extra info in the body of the method :/

zachmargolis replied:
I decided to retry this in d4efd77f6a155d8933468e18c73fe8513a5ddb3d and removed it by changing the analytics event to use default keyword arguments

aduth replied:
Ah, I think I missed that the place where the parameter was passed was within `analytics_events.rb` itself.

zachmargolis commented:
Ok update! The new plan here is that I added the infra needed to error on unknown params (and even document string keys), but plumbing those through was a lot of work and very tedious so I added an escape valve:

```ruby
allowed_extra_analytics: [:*]
```

and am annotating failing specs with those so we can address documentation in a more piecemeal fashion
